### PR TITLE
feat: initial implementation of V2.Departure and Departure widgets

### PR DESCRIPTION
**Pull request by handorff**
**Asana tasks**:
- [Define Departures backend widgets](https://app.asana.com/0/1185117109217413/1200233815812787)
- [Create V2.Departure](https://app.asana.com/0/1185117109217413/1200233815812791)

### CODE REVIEW

**Review by jzimbel-mbta — COMMENTED** [review]
What an improvement! I have a few comments but nothing major. The number of comments is just due to the size of the review.

File: lib/screens/v2/widget_instance/departures.ex [hunk 1]
```diff
@@ -1,41 +1,261 @@
...
+
+        _ ->
+          %{time: serialized_time, scheduled_time: serialized_scheduled_time}
+      end
+    end
+  end
+
+  defp serialize_timestamp(departure_time) do
+    {:ok, local_time} = DateTime.shift_zone(departure_time, "America/New_York")
+    hour = 1 + rem(local_time.hour - 1, 12)
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/departures.ex:229]: I think if you want hour = 12 when local_time.hour is 0, you can use [`Integer.mod/2`](https://hexdocs.pm/elixir/Integer.html#mod/2).
```ex
iex> 1 + Integer.mod(0 - 1, 12)
#=> 12
```

File: lib/screens/v2/departure.ex [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    if opts[:include_schedules] do
+      fetch_predictions_and_schedules(params)
+    else
+      fetch_predictions_only(params)
+    end
+  end
+
+  def fetch_predictions_and_schedules(params) do
+    with {:ok, predictions} <- Prediction.fetch(params),
+         {:ok, schedules} <- Schedule.fetch(params) do
```
[the remark this answers is not in the document]
Comment (handorff) [reply lib/screens/v2/departure.ex:29]: We'd need to do something when we deprecate the old `Departure` module, but I think it'd be OK to pick that up separately later.

File: lib/screens/v2/departure.ex [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+  alias Screens.Predictions.Prediction
+  alias Screens.Routes.Route
+  alias Screens.Schedules.Schedule
+  alias Screens.Stops.Stop
+  alias Screens.Trips.Trip
+  alias Screens.Vehicles.Vehicle
+
+  @type t :: %__MODULE__{
+          prediction: Screens.Predictions.Prediction.t() | nil,
+          schedule: Screens.Schedules.Schedule.t() | nil
```
[the remark this answers is not in the document]
Comment (handorff) [reply lib/screens/v2/departure.ex:13]: No, and we don't handle this case in the accessor functions.

File: lib/screens/v2/departure.ex [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    predicted_trip_ids =
+      predictions
+      |> Enum.reject(&is_nil(&1.trip))
+      |> Enum.map(& &1.trip.id)
+      |> Enum.reject(&is_nil/1)
+      |> Enum.into(MapSet.new())
+
+    schedules_by_trip_id =
+      schedules
+      |> Enum.map(fn %{trip: %{id: trip_id}} = s -> {trip_id, s} end)
```
[the remark this answers is not in the document]
Comment (handorff) [reply lib/screens/v2/departure.ex:123]: I don't think this is possible. `trip_id` is required for `stop_times.txt` in GTFS.

[the remark this answers is not in the document]
Comment (handorff) [reply lib/screens/v2/departure.ex:123]: (But I agree with your general point, we do have to check both of those a lot.)

File: lib/screens/v2/widget_instance/departures.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (handorff) [reply lib/screens/v2/widget_instance/departures.ex:229]: Thanks! That was indeed my intention, and I didn't realize that `rem(-1, 12)` was -1. I updated the timestamp test to cover this case.

File: test/screens/v2/departure_test.exs [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+    test "returns schedule route short_name when no prediction is present" do
+      schedule = %Schedule{route: %Route{id: "214216", short_name: "214/216"}}
+      departure = %Departure{prediction: nil, schedule: schedule}
+
+      assert "214/216" == Departure.route_name(departure)
+    end
+  end
+
+  describe "route_type/1" do
```
Comment (jzimbel-mbta) [inline test/screens/v2/departure_test.exs:197]: Could you update these tests to use atom route types?

**Review by jzimbel-mbta — APPROVED** [review]
One small thing, otherwise good to go!

**Pull request merged**